internal/cache/lrucache/lrucache.go

```diff
@@ -0,0 +1,294 @@
+		case entry:
+		default:
+			panic(fmt.Sprintf("Unexpected element type: %v", reflect.TypeOf(e.Value)))
+		}
+	}
+
+	// INVARIANT: For each k, v: v.Value.(entry).Key == k
+	// INVARIANT: Contains all and only the elements of entries
+	if c.entries.Len() != len(c.index) {
+		panic(fmt.Sprintf(
+			"Length mismatch: %v vs. %v",
+			c.entries.Len(),
+			len(c.index)))
+	}
+
+	for e := c.entries.Front(); e != nil; e = e.Next() {
+		if c.index[e.Value.(entry).Key] != e {
+			panic(fmt.Sprintf("Mismatch for key %v", e.Value.(entry).Key))
+		}
+	}
+}
+
+func (c *Cache) evictOne() ValueType {
+	e := c.entries.Back()
+	key := e.Value.(entry).Key
+
+	evictedEntry := e.Value.(entry).Value
+	c.currentSize -= evictedEntry.Size()
+
+	c.entries.Remove(e)
+	delete(c.index, key)
+
+	return evictedEntry
+}
+
+////////////////////////////////////////////////////////////////////////
+// Cache interface
+////////////////////////////////////////////////////////////////////////
+
+// Insert the supplied value into the cache, overwriting any previous entry for
+// the given key. The value must be non-nil.
+// Also returns a slice of ValueType evicted by the new inserted entry.
+// LOCK_EXCLUDED(c.mu)
+func (c *Cache) Insert(
+	key string,
+	value ValueType) []ValueType {
+	if value == nil {
+		panic("nil values are not supported")
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	e, ok := c.index[key]
+	if ok {
+		// Update an entry if already exist.
+		c.currentSize -= e.Value.(entry).Value.Size()
+		c.currentSize += value.Size()
+		e.Value = entry{key, value}
+		c.entries.MoveToFront(e)
+	} else {
+		// Add the entry if already doesn't exist.
+		e := c.entries.PushFront(entry{key, value})
+		c.index[key] = e
+		c.currentSize += value.Size()
+	}
+
+	var evictedValues []ValueType
+	// Evict until we're at or below maxSize.
+	for c.currentSize > c.maxSize {
+		evictedValues = append(evictedValues, c.evictOne())
+	}
+
+	return evictedValues
+}
+
+// Erase any entry for the supplied key, also returns the value of erased key.
+// LOCK_EXCLUDED(c.mu)
+func (c *Cache) Erase(key string) (value ValueType) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	e := c.index[key]
+	if e == nil {
+		return
+	}
+
+	deletedEntry := e.Value.(entry).Value
+	c.currentSize -= value.Size()
+
+	delete(c.index, key)
+	c.entries.Remove(e)
+
+	return deletedEntry
+}
+
+// LookUp a previously-inserted value for the given key. Return nil if no
+// value is present.
+// LOCK_EXCLUDED(c.mu)
+func (c *Cache) LookUp(key string) (value ValueType) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Consult the index.
+	e := c.index[key]
+	if e == nil {
+		return
+	}
+
+	// This is now the most recently used entry.
+	c.entries.MoveToFront(e)
+
+	// Return the value.
+	return e.Value.(entry).Value
+}
+
+////////////////////////////////////////////////////////////////////////
+// Gob encoding
+////////////////////////////////////////////////////////////////////////
+
+func (c *Cache) GobEncode() (b []byte, err error) {
```

Comment:
Can we remove this to reduce code if not required ? We can include whenever there is a requirement.